lxd/auth/driver_tls.go

```diff
@@ -1,65 +1,427 @@
+// UserAccess is not supported.
+func (a *tls) UserAccess(ctx context.Context, username string) (*UserAccess, error) {
+	return nil, ErrNotSupported
 }
 
-// RenameProject is a no-op. It notifies the authorization service that a project has been renamed.
-func (a *tls) RenameProject(projectID int64, newName string) error {
-	return nil
-}
+// UserHasPermission checks whether the requestor has a specific permission on a project.
+func (a *tls) UserHasPermission(ctx context.Context, r *http.Request, object string, entitlement Entitlement) bool {
+	val := r.Context().Value(request.CtxAccess)
+	if val == nil {
+		return false
+	}
 
-// StopStatusCheck is a no-op.
-func (a *tls) StopStatusCheck() {
-}
+	ua, ok := val.(*UserAccess)
+	if !ok {
+		return false
+	}
 
-func (a *tls) UserAccess(username string) (*UserAccess, error) {
-	return &UserAccess{Admin: true}, nil
-}
+	if ua.Admin {
+		return true
+	}
 
-// UserIsAdmin checks whether the requestor is a global admin.
-func (a *tls) UserIsAdmin(r *http.Request) bool {
-	val := r.Context().Value(request.CtxAccess)
-	if val == nil {
+	permission, projectName, allow, err := a.relationToPermission(object, entitlement)
+	if err != nil {
+		a.logger.Error("Failed to check permission for object", logger.Ctx{"object": object, "entitlement": entitlement, "error": err})
 		return false
 	}
 
-	ua := val.(*UserAccess)
-	return ua.Admin
+	if allow {
+		return true
+	}
+
+	return shared.ValueInSlice(string(permission), ua.Projects[projectName])
 }
 
-// UserHasPermission checks whether the requestor has a specific permission on a project.
-func (a *tls) UserHasPermission(r *http.Request, projectName string, permission string) bool {
+// GetPermissionChecker returns a function that can be used to check whether a user has the required entitlement on an authorization object.
+func (a *tls) GetPermissionChecker(ctx context.Context, r *http.Request, entitlement Entitlement, objectType ObjectType) (func(object string) bool, error) {
+	allowFunc := func(b bool) func(string) bool {
+		return func(string) bool {
+			return b
+		}
+	}
+
 	val := r.Context().Value(request.CtxAccess)
 	if val == nil {
-		return false
+		return allowFunc(false), nil
+	}
+
+	ua, ok := val.(*UserAccess)
+	if !ok {
+		return allowFunc(false), nil
 	}
 
-	ua := val.(*UserAccess)
 	if ua.Admin {
-		return true
+		return allowFunc(true), nil
 	}
 
-	return shared.ValueInSlice(permission, ua.Projects[projectName])
+	return func(object string) bool {
+		permission, projectName, allow, err := a.relationToPermission(object, entitlement)
+		if err != nil {
+			a.logger.Error("Failed to check permission for object", logger.Ctx{"object": object, "entitlement": entitlement, "error": err})
+			return false
+		}
+
+		if allow {
+			return true
+		}
+
+		return shared.ValueInSlice(string(permission), ua.Projects[projectName])
+	}, nil
+}
+
+func (a *tls) relationToPermission(object string, entitlement Entitlement) (Permission, string, bool, error) {
```

Comment:
In this function, is there a way to make it less maintenance heavy in the future?
What I mean is that TLS auth is pretty coarse, so do we need to define a mapping 1:1 between every object type and entitlement or can we make this more general so it doesn't need to be updated for every new object and entitlement?

Reply:
I don't think it's going to be possible to generalise this unfortunately. The good news (as we discussed earlier) is that if we can't match the object and entitlement then this function will return an error, which will result in a `403 Forbidden`. We should be able to catch this easily in tests and if one slips by it'll be an easy fix.

Reply:
I'm not sure what you mean. `relationToPermission` does extract the project from the object. This is used in the `UserHasPermission` method along with the permission to check the `UserAccess` struct for the permission on that project. The `UserAccess` struct is set in the request context using the users certificate before any authorization calls are made.